aws/resource_aws_api_gateway_documentation_version_test.go

```diff
@@ -0,0 +1,266 @@
+	"github.com/hashicorp/terraform/helper/acctest"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAWSAPIGatewayDocumentationVersion_basic(t *testing.T) {
+	var conf apigateway.DocumentationVersion
+
+	rString := acctest.RandString(8)
+	version := fmt.Sprintf("tf_acc_version_%s", rString)
+	apiName := fmt.Sprintf("tf_acc_api_doc_version_basic_%s", rString)
+
+	resourceName := "aws_api_gateway_documentation_version.test"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSAPIGatewayDocumentationVersionDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSAPIGatewayDocumentationVersionBasicConfig(version, apiName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSAPIGatewayDocumentationVersionExists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "version", version),
+					resource.TestCheckResourceAttrSet(resourceName, "rest_api_id"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSAPIGatewayDocumentationVersion_allFields(t *testing.T) {
+	var conf apigateway.DocumentationVersion
+
+	rString := acctest.RandString(8)
+	version := fmt.Sprintf("tf_acc_version_%s", rString)
+	apiName := fmt.Sprintf("tf_acc_api_doc_version_method_%s", rString)
+	stageName := fmt.Sprintf("tf_acc_stage_%s", rString)
+	description := fmt.Sprintf("Tf Acc Test description %s", rString)
+	uDescription := fmt.Sprintf("Tf Acc Test description updated %s", rString)
+
+	resourceName := "aws_api_gateway_documentation_version.test"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSAPIGatewayDocumentationVersionDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccAWSAPIGatewayDocumentationVersionAllFieldsConfig(version, apiName, stageName, description),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSAPIGatewayDocumentationVersionExists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "version", version),
+					resource.TestCheckResourceAttr(resourceName, "description", description),
+					resource.TestCheckResourceAttrSet(resourceName, "rest_api_id"),
+				),
+			},
+			resource.TestStep{
+				Config: testAccAWSAPIGatewayDocumentationVersionAllFieldsConfig(version, apiName, stageName, uDescription),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSAPIGatewayDocumentationVersionExists(resourceName, &conf),
+					resource.TestCheckResourceAttr(resourceName, "version", version),
+					resource.TestCheckResourceAttr(resourceName, "description", uDescription),
+					resource.TestCheckResourceAttrSet(resourceName, "rest_api_id"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSAPIGatewayDocumentationVersion_importBasic(t *testing.T) {
```

Comment:
Nitpick: Something I've been trying to remember to do is include `ImportState` `TestStep`s with each of the regular tests. We get the extra benefit of checking the additional attributes associated with each of the test cases and don't need explicit import test cases (which generally get neglected unless there's a 🐛  - too late!).

Reply:
I have no objections against adding more import tests with different configurations, but I don't think it's a good idea to add `ImportState` to existing tests which aren't called `_import`.

If we break import (for any reason) and nothing else I expect `_import` tests to fail, but no other tests, and vice versa, ideally. While in reality this is may not be always true I think we should aim for such isolation so it's easier to debug test failures as they appear.

There's benefit to isolation, I think - to keep us sane and focused 😉

Reply:
I added a new test - `TestAccAWSAPIGatewayDocumentationVersion_importAllFields` which should address this concern, hopefully? 😃

Reply:
I don't mean to harp on this resource/test specifically, but here's where I'm coming from, using this resource's testing as an example. I'm not asking you to change this one, but I think its something worth thinking about. 😄 

* We now have two of these import tests, duplicating the regular ones except a new `TestStep`. This means the testing is going to take roughly twice as long. Not the biggest deal, but even running parallel these all add up. e.g.

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSAPIGatewayDocumentationVersion'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAPIGatewayDocumentationVersion -timeout 120m
=== RUN   TestAccAWSAPIGatewayDocumentationVersion_basic
--- PASS: TestAccAWSAPIGatewayDocumentationVersion_basic (14.69s)
=== RUN   TestAccAWSAPIGatewayDocumentationVersion_allFields
--- PASS: TestAccAWSAPIGatewayDocumentationVersion_allFields (57.37s)
=== RUN   TestAccAWSAPIGatewayDocumentationVersion_importBasic
--- PASS: TestAccAWSAPIGatewayDocumentationVersion_importBasic (16.63s)
=== RUN   TestAccAWSAPIGatewayDocumentationVersion_importAllFields
--- PASS: TestAccAWSAPIGatewayDocumentationVersion_importAllFields (43.97s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	132.701s
```

* The "regular" part (first `TestStep`) of these `_importX` tests can also fail because of other resources. There are plenty of examples of these in the EC2 or KMS daily acceptance testing. So if you're looking at just the test name and not the actual error message (we're looking at the error message right? 😉 ), you can be misled already.

* I think `TestStep.ImportStateVerify` seems to do a pretty good job of calling out that its an import problem by the error message.

```go
// Comment out in resourceAwsApiGatewayDocumentationVersionRead
// d.Set("description", version.Description)

// Add in resourceAwsApiGatewayDocumentationVersionCreate
d.Set("description", d.Get("description").(string))

// Add to TestAccAWSAPIGatewayDocumentationVersion_allFields
			resource.TestStep{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```

It returns:
```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSAPIGatewayDocumentationVersion_allFields'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSAPIGatewayDocumentationVersion_allFields -timeout 120m
=== RUN   TestAccAWSAPIGatewayDocumentationVersion_allFields
--- FAIL: TestAccAWSAPIGatewayDocumentationVersion_allFields (34.59s)
	testing.go:513: Step 2 error: ImportStateVerify attributes not equivalent. Difference is shown below. Top is actual, bottom is expected.

		(map[string]string) {
		}


		(map[string]string) (len=1) {
		 (string) (len=11) "description": (string) (len=40) "Tf Acc Test description updated xjsh6nsw"
		}

FAIL
exit status 1
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	34.628s
```

Combined with maybe a little bit better naming of the tests (e.g. `_description` in this case), we'll know from the test name its a `description` problem and from the error message its an import problem. 

Hope this makes sense! 🍻  I guess mentally I am treating an attribute's import just a "feature" of an attribute, which should be logically tested with the other CRUD actions on it.

Reply:
> mentally I am treating an attribute's import just a "feature" of an attribute, which should be logically tested with the other CRUD actions on it.

That's an interesting POV. I never thought about that like this. We can discuss that on Slack with other folks later. 👌